File: app/src/main/java/com/esoxjem/movieguide/listing/MoviesListingInteractorImpl.java
package com.esoxjem.movieguide.listing;

import com.esoxjem.movieguide.Movie;
import com.esoxjem.movieguide.MoviesWraper;
import com.esoxjem.movieguide.favorites.FavoritesInteractor;
import com.esoxjem.movieguide.listing.sorting.SortType;
import com.esoxjem.movieguide.listing.sorting.SortingOptionStore;
import com.esoxjem.movieguide.network.TmdbWebService;

import java.util.List;

import io.reactivex.Observable;

/**
 * @author arun
 */
class MoviesListingInteractorImpl implements MoviesListingInteractor {
    private FavoritesInteractor favoritesInteractor;
    private TmdbWebService tmdbWebService;
    private SortingOptionStore sortingOptionStore;

    MoviesListingInteractorImpl(FavoritesInteractor favoritesInteractor,
                                TmdbWebService tmdbWebService, SortingOptionStore store) {
        this.favoritesInteractor = favoritesInteractor;
        this.tmdbWebService = tmdbWebService;
        sortingOptionStore = store;
    }

    @Override
    public Observable<List<Movie>> fetchMovies() {
        int selectedOption = sortingOptionStore.getSelectedOption();
        if (selectedOption == SortType.MOST_POPULAR.getValue()) {
            return tmdbWebService.popularMovies().map(MoviesWraper::getMovieList);
        } else if (selectedOption == SortType.HIGHEST_RATED.getValue()) {
            return tmdbWebService.highestRatedMovies().map(MoviesWraper::getMovieList);
        } else {
            return Observable.just(favoritesInteractor.getFavorites());
        }
    }

}

File: app/src/main/java/com/esoxjem/movieguide/movies/listing/MoviesListingInteractorImpl.kt
package com.esoxjem.movieguide.movies.listing

import com.esoxjem.movieguide.movies.entities.Movie
import com.esoxjem.movieguide.movies.favorites.FavoritesInteractor
import com.esoxjem.movieguide.movies.sorting.SortType
import com.esoxjem.movieguide.movies.sorting.SortingOptionStore
import com.esoxjem.movieguide.network.TmdbWebService
import io.reactivex.Observable
import java.text.SimpleDateFormat
import java.util.*

/**
 * @author arunsasidharan
 */
internal class MoviesListingInteractorImpl(private val favoritesInteractor: FavoritesInteractor,
                                           private val tmdbWebService: TmdbWebService, private val sortingOptionStore: SortingOptionStore) : MoviesListingInteractor {
    private val dateFormat = SimpleDateFormat("yyyy-MM-dd")


    override val isPaginationSupported: Boolean
        get() {
            val selectedOption = sortingOptionStore.getSelectedOption()
            return selectedOption != SortType.FAVORITES.value
        }

    override fun fetchMovies(page: Int): Observable<List<Movie>> {
        val selectedOption = sortingOptionStore.getSelectedOption()
        return when (selectedOption) {
            SortType.MOST_POPULAR.value -> tmdbWebService.popularMovies(page).map { it.movieList }

            SortType.HIGHEST_RATED.value -> tmdbWebService.highestRatedMovies(page).map { it.movieList }

            SortType.NEWEST.value -> {
                val cal = Calendar.getInstance()
                val maxReleaseDate = dateFormat.format(cal.time)
                tmdbWebService.newestMovies(maxReleaseDate, 50).map { it.movieList }
            }
            else -> Observable.just(favoritesInteractor.getFavorites())
        }
    }
}
